ftl/common/build_layer.py

```diff
@@ -0,0 +1,155 @@
+    layer.
+    """
+
+    __metaclass__ = abc.ABCMeta  # For enforcing that methods are overriden.
+
+    def __init__(self):
+        self._img = None
+        self._lyr = None
+        self._diff_id = None
+        self._overrides = None
+
+    @abc.abstractmethod
+    def GetImage(self):
+        """Get the.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetLayer(self):
+        """Get the.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetDiffId(self):
+        """Get the.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetCacheKeyRaw(self):
+        """Synthesizes the application layer from the context.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetCacheKey(self):
+        """Synthesizes the application layer from the context.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetOverrides(self):
+        """Synthesizes the application layer from the context.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def BuildLayer(self):
+        """Synthesizes the application layer from the context.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+# TODO(aaron-prindle) create custom error type
+
+class CacheLayerImage(BaseLayerImage):
+
+    def GetCacheRaw(self):
+        pass
+
+    def BuildLayer(self):
+        pass
+
+    def GetCacheKey(self):
+        return docker_digest.SHA256(self.GetCacheKeyRaw())
+
+    def GetImage(self):
+        if self._lyr is None:
+            raise Exception('error: layer image was not built yet so \
+                             image cannot be accessed')
+        if self._img is None:
+            self._img = tar_to_dockerimage.TarDockerImage(self._lyr)
+        return self._img
+
+    def GetDiffId(self):
+        if self._diff_id is None:
+            raise Exception('error: layer image was not built yet so \
+                             diff_id cannot be accessed')
+        return self._diff_id
+
+    def GetLayer(self):
+        if self._lyr is None:
+            raise Exception('error: layer image was not built yet so \
+                             layer cannot be accessed')
+        return self._lyr
+
+    def GetOverrides(self):
+        if self._overrides is None:
+            raise Exception('error: layer image was not built yet so \
+                             overrides cannot be accessed')
+        return self._overrides
+
+    def CacheLayerImage(self, cash, base, namespace):
+        cash.Store(base, namespace, self.GetCacheKey(),
+                        self.GetImage())
+
+    def IsCached(self, cash, base, namespace, cache_key=None):
```

Comment:
should this be moved to the cache object instead?

Reply:
Yeah, looking more at this I think it would make sense for the caching logic to live in a cache class, not on the layer itself. The layer can know about it's own cache key and any cache key's it's dependent on, but not necessarily how to store itself or check the cache.

Reply:
These methods can all move, I just added them here